src/metabase/pulse.clj

```diff
@@ -80,20 +80,48 @@
     (compare (:row dashcard-1) (:row dashcard-2))
     (compare (:col dashcard-1) (:col dashcard-2))))
 
+(defn- virtual-card-of-type?
+  "Check if dashcard is a virtual with type `ttype`.
+
+  There are currently 3 types of virtual card: text, action, link."
+  [dashcard ttype]
+  (and (map? dashcard)
+       (= ttype (get-in dashcard [:visualization_settings :virtual_card :display]))))
+
+(defn- dashcard->content
+  "Given a dashcard returns its content based on its type."
+  [dashcard {pulse-creator-id :creator_id, :as pulse} dashboard]
+  (cond
+    (:card_id dashcard)
+    (let [parameters (merge-default-values (params/parameters pulse dashboard))]
+      (execute-dashboard-subscription-card pulse-creator-id dashboard dashcard (:card_id dashcard) parameters))
+
+    ;; actions
+    (virtual-card-of-type? dashcard "action")
+    nil
+
+    ;; link cards
+    (virtual-card-of-type? dashcard "link")
+    nil
```

Comment:
We'll add support for link cards [PR](https://github.com/metabase/metabase/pull/28528), so this part is a bit explicit to use as template for the follow up PR.